esp-hal/src/sdm.rs

```diff
@@ -0,0 +1,296 @@
+//! # Sigma-Delta modulation (SDM)
+//!
+//! ## Overview
+//!
+//! Almost all ESP SoCs has a second-order sigma-delta modulator, which
+//! can generate independent PDM pulses to multiple channels. Please refer
+//! to the TRM to check how many hardware channels are available.
+//!
+//! Delta-sigma modulation converts an analog voltage signal into a pulse
+//! frequency, or pulse density, which can be understood as pulse-density
+//! modulation (PDM) (refer to [Delta-sigma modulation on Wikipedia](https://en.wikipedia.org/wiki/Delta-sigma_modulation)).
+//!
+//! Typically, a Sigma-Delta modulated channel can be used in scenarios like:
+//!
+//! - LED dimming
+//! - Simple DAC (8-bit), with the help of an active RC low-pass filter
+//! - Class D amplifier, with the help of a half-bridge or full-bridge circuit
+//!   plus an LC low-pass filter
+//!
+//! ## Configuration
+//!
+//! After creating [`Sdm`] instance you should connect individual channels to
+//! GPIO outputs. Also you need set modulation frequency.
+//!
+//! ## Usage
+//!
+//! Connected channels accepts pulse density in range -128..127.
+
+use core::marker::PhantomData;
+
+use fugit::HertzU32;
+
+use crate::{
+    clock::Clocks,
+    gpio::{OutputPin, OutputSignal},
+    peripheral::{Peripheral, PeripheralRef},
+    peripherals::GPIO_SD,
+    private,
+};
+
+/// Sigma-Delta modulation peripheral driver.
+pub struct Sdm<'d> {
+    /// Channel 0
+    pub channel0: ChannelRef<'d, 0>,
+
+    /// Channel 1
+    pub channel1: ChannelRef<'d, 1>,
+
+    /// Channel 2
+    pub channel2: ChannelRef<'d, 2>,
+
+    /// Channel 3
+    pub channel3: ChannelRef<'d, 3>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 4
+    pub channel4: ChannelRef<'d, 4>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 5
+    pub channel5: ChannelRef<'d, 5>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 6
+    pub channel6: ChannelRef<'d, 6>,
+
+    #[cfg(any(esp32, esp32s2, esp32s3))]
+    /// Channel 7
+    pub channel7: ChannelRef<'d, 7>,
+}
+
+/// Channel errors
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// Prescale out of range
+    PrescaleRange,
+}
+
+impl<'d> Drop for Sdm<'d> {
+    fn drop(&mut self) {
+        GPIO_SD::enable_clock(false);
+    }
+}
```

Comment:
Hmm, having a `Drop` implementation here means that the channels can't be sent off to other threads/tasks/cores/interrupts.

I'm wondering if there are any downsides of leaving it out.

Reply:
Yes, the channels themselves should track how many are still enabled (with an AtomicUsize for example) and disable the clocks if all are gone. The "collection" peripheral shouldn't be Drop because the usual intent is to move out of it. We are putting work into establishing these patterns in this dev cycle, along with implementing Drop for peripherals so that they clean up after themselves. I'm not sure it's worth doing similar work in parallel until we arrive at something we like.

Reply:
Personally I'd like an explicit drop. The user can recreate the "collection" peripheral from the parts that were split, and then call some `release()` method to safely get the peripheral back (which would disable clocks, reset, etc.).
Only two issues are it doesn't quite play nicely with the PeripheralRef pattern (but it works better for `static`s) and it's somewhat tedious (though I reckon few applications will actually want to release their drivers anyway).

Reply:
@Dominaezzz When I tried to share LEDC channel with interrupt handler I cannot do that. Seems current design lacks support for such usage scenarios.

@bugadani I think that ability to deinit periphery is matter especially for applications which needs power management.

@Dominaezzz In my opinion explicit drop via API looks like a low level thing and we should avoid it as much as possible.